delegate-app/Main.hs

```diff
@@ -1,49 +1,179 @@
+delegateServerApp pingSecs reqq broadcast pending = do
+  connection <- liftIO $ acceptRequest pending
+
+  trace FrontendConnected
+  tracer <- askTracer
+
+  liftIO $ do
+    broadcast' <- atomically $ dupTChan broadcast
+
+    let encodeSend :: forall a. ToJSON a => a -> IO ()
+        encodeSend = sendTextData connection . encode
+
+        receive = forever $
+          runWithTracer' tracer $ do
+            inp <- liftIO $ receiveData connection
+            case eitherDecode @FrontendRequest inp of
+              Left msg -> do
+                let err = FrontendNoParse msg
+                trace (DelegateError err)
+                liftIO $ encodeSend err
+              Right req -> do
+                trace $ FrontendInput req
+                liftIO . atomically . writeTQueue reqq $ FrontendRequest req
+
+        send =
+          forever $
+            atomically (readTChan broadcast')
+              >>= encodeSend
+
+    withPingThread connection pingSecs (pure ()) $ race_ receive send -- TODO: can we do this
 
-eventsProducer :: Bool -> DelegateRunnerT IO (Maybe DelegateEvent)
-eventsProducer thisIsFirstRun =
-  if thisIsFirstRun
-    then return $ Just Start
-    else return Nothing
-
-consumer :: Maybe DelegateInput -> DelegateRunnerT IO ()
-consumer mInput = case mInput of
-  Just input -> do
-    delegateResponse <- delegateStep input
-    -- FIXME: send response to client
-    liftIO $
-      putStrLn $
-        "Delegate responses for input: " <> show delegateResponse
-  Nothing -> return ()
-
-delegateTick :: Int
-delegateTick = 1_000
+{- | create a delegate runner that will listen for incoming events and output outgoing events
+   accordingly
+-}
+mkRunner ::
+  -- | the time in milliseconds that the runner sleeps between acts
+  Int ->
+  -- | the queue of incoming messages
+  TQueue DelegateInput ->
+  -- | the broadcast queue of outgoing messages (write only)
+  TChan DelegateResponse ->
+  DelegateRunnerT (DelegateTracerT IO) ()
+-- FIXME: we need to abort at some point but this doesn't seem
+-- to be implemented yet so we just go on
+mkRunner tick reqq broadcast = forever $ do
+  let encodeBroadcast responses = do
+        liftIO . atomically $ traverse_ (writeTChan broadcast) responses
+        lift $ traverse_ (trace . DelegateOutput) responses
+
+  events <- liftIO . atomically $ flushTQueue reqq
+  traverse_ (delegateStep >=> encodeBroadcast) events
+
+  liftIO $ threadDelay tick
+
+-- | run a delegate server
+runDelegateServer ::
+  -- | the configuration of the delegate server
+  DelegateServerConfig ->
+  TracerT DelegateServerLog IO ()
+runDelegateServer conf = do
+  trace (Started $ dlgt'port conf)
+
+  tracer <- askTracer
+
+  delegateInputChan <- liftIO . atomically $ do
+    q <- newTQueue
+    writeTQueue q (DelegateEvent Start)
+    pure q
+  delegateBroadCastChan <- liftIO . atomically $ newTChan
+
+  _ <-
+    liftIO . forkIO $
+      runServer (show $ dlgt'host conf) (fromIntegral $ dlgt'port conf) $
```

Comment:
> fromIntegral

Why not just store it in that type already in config?

Reply:
Because it's basically untyped. The proper server I had before took proper types. This doesn't. If you wish I can change it. Be aware that we don't validate IPs then anymore (which we do now by parsing IPs)

Reply:
Ah, there is some specialized type in config, understand now, you right.

Reply:
Ah, there is some specific port type in config, understand now, you right.